gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -599,6 +721,671 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
 }
 
+/************************************************************************/
+/*                          WriteFITSInfo()                          */
+/************************************************************************/
+
+void FITSDataset::WriteFITSInfo()
+
+{
+  int status = 0;
+
+  const double PI = std::atan(1.0)*4;
+  const double DEG2RAD = PI / 180.;
+
+  double falseEast = 0;
+  double falseNorth = 0;
+
+  double cfactor, mres, mapres, UpperLeftCornerX, UpperLeftCornerY;
+  double crpix1, crpix2;
+
+/* -------------------------------------------------------------------- */
+/*      Write out projection definition.                                */
+/* -------------------------------------------------------------------- */
+    const bool bHasProjection = !oSRS.IsEmpty();
+    if( bHasProjection )
+    {
+
+        // Set according to coordinate system (thanks to Trent Hare - USGS)
+
+        char object[16];
+        char ctype1[9], ctype2[9];
+        
+        const char* target = oSRS.GetAttrValue("DATUM",0);
+        if ( strstr(target, "Moon") ) {
+          strncpy(object, "Moon",5);
+          strncpy(ctype1, "SE", 3);
+          strncpy(ctype2, "SE", 3);
+        } else if ( strstr(target, "Mercury") ) {
+          strncpy(object, "Mercury", 8);      
+          strncpy(ctype1, "ME", 3);
+          strncpy(ctype2, "ME", 3);
+        } else if ( strstr(target, "Venus") ) {
+          strncpy(object, "Venus", 6);
+          strncpy(ctype1, "VE", 3);
+          strncpy(ctype2, "VE", 3);
+        } else if ( strstr(target, "Mars") ) {
+          strncpy(object, "Mars", 5);
+          strncpy(ctype1, "MA", 3);
+          strncpy(ctype2, "MA", 3);
+        } else if ( strstr(target, "Jupiter") ) {
+          strncpy(object, "Jupiter", 8);
+          strncpy(ctype1, "JU", 3);
+          strncpy(ctype2, "JU", 3);
+        } else if ( strstr(target, "Saturn") ) {
+          strncpy(object, "Saturn", 7);
+          strncpy(ctype1, "SA", 3);
+          strncpy(ctype2, "SA", 3);
+        } else if ( strstr(target, "Uranus") ) {
+          strncpy(object, "Uranus", 7);
+          strncpy(ctype1, "UR", 3);
+          strncpy(ctype2, "UR", 3);
+        } else if ( strstr(target, "Neptune") ) {
+          strncpy(object, "Neptune", 8);
+          strncpy(ctype1, "NE", 3);
+          strncpy(ctype2, "NE", 3);
+        }
+
+        fits_update_key( hFITS, TSTRING, "OBJECT", &object, nullptr, &status);
+
+        double aradius = oSRS.GetSemiMajor();
+        double bradius = aradius;
+        double cradius = oSRS.GetSemiMinor();
+
+        cfactor = aradius * DEG2RAD;
+
+        fits_update_key( hFITS, TDOUBLE, "A_RADIUS", &aradius, nullptr, &status);
+        fits_update_key( hFITS, TDOUBLE, "B_RADIUS", &bradius, nullptr, &status);
+        fits_update_key( hFITS, TDOUBLE, "C_RADIUS", &cradius, nullptr, &status);
+
+        const char* un = oSRS.GetAttrValue("UNIT",0);
+
+        strcat(ctype1, "LN-");
+        strcat(ctype2, "LT-"); 
```

Comment:
It starts to get really unclear that you have enough space in ctype1 and ctype2 for this.  std::string would eliminate this issue, or you could just set aside enough memory so that it couldn't reasonable be an issue.  This had me going back to check byte counts.

Reply:
CTYPE1 and CTYPE2 are by default 8 chars in FITS, anyway you're right std::string is a better solution. I was probably afraid of passing from string to char when updating FITS keywords with CFITSIO... It cames out simpler than I thought... :)
Fixed in 41699ae24b983381a1a63715fbbae71c442059d8